### ISSUE TITLE
Multithreaded processing

### ISSUE THREAD

**Issue by Interspatial** [CPJKU/madmom#440]
Hi,
I'm trying to process a batch of audio signals by first creating a Signal for each of them and running them in parallel. However, I'm getting an error: 
```
AttributeError: 'Signal' object has no attribute 'sample_rate'
```

Code:
```
fsig_proc = FramedSignalProcessor([...[)
spec_proc = LogarithmicFilteredSpectrogramProcessor([...])
processor = SequentialProcessor([fsig_proc, spec_proc])

signals = []
for signal in y:
    signals.append(Signal(np.array(signal), sample_rate=kwargs['fs']))

specs = Parallel(n_jobs=kwargs['num_threads'])(
             delayed(processor)(signal) for signal in signals)
```

Thanks

**Comment by fdlm** [comment]
Works for me with the latest madmom version. Which version do you use? Have you tried installing from source?

**Comment by Interspatial** [comment]
I'm running on Ubuntu with madmom ver. 0.16.1. It has to do something with threading since while it works fine with ```n_jobs=1```, increasing ```n_jobs``` causes this error.

**Comment by superbock** [comment]
Unfortunately, the code you posted is incomplete, e.g. what's inside `[...]`, is `Parallel` `joblib.Parallel`, etc.? Thus I can only have a guess. Since madmom uses multiprocessing as well (set via `num_threads`), it's quite likely that this interferes with your Parallel and has unwanted side effects. Can you post a minimal working example where only changing one parameter leads to the error?

**Comment by laplaceyang** [comment]
same problem for me, I use python 3.6.5
```
class MadmomProcessor(object):
    __instance = {}
    def __init__(self, frame_size, num_bands):
        frames = FramedSignalProcessor(frame_size=frame_size, fps=100)
        stft = ShortTimeFourierTransformProcessor()  # caching FFT window
        filt = FilteredSpectrogramProcessor(
            num_bands=num_bands, fmin=30, fmax=17000, norm_filters=True)
        spec = LogarithmicSpectrogramProcessor(mul=1, add=1)
        diff = SpectrogramDifferenceProcessor(
            diff_ratio=0.5, positive_diffs=True, stack_diffs=np.hstack)
        # process each frame size with spec and diff sequentially
        self.processor = SequentialProcessor((frames, stft, filt, spec, diff))

    def process(self, signal):
        return self.processor(signal)

    @classmethod
    def Process(cls, signal, frame_size, num_bands):
        key = "%s-%s" % (frame_size, num_bands)
        if key not in cls.__instance:
            cls.__instance[key] = MadmomProcessor(frame_size, num_bands)
        return cls.__instance[key].process(signal)


def madmom_feature(signal, frame_size, num_bands):
    return MadmomProcessor.Process(signal, frame_size, num_bands)

def madmom_feature_1024_3(signal):
    print(type(signal).__name__)
    return madmom_feature(signal, 1024, 3)

class SignalLoader(object):
    __instance = None
    def __init__(self):
        self.sig = SignalProcessor(num_channels=1, sample_rate=SAMPLE_RATE, dtype=np.float32)

    def process(self, data):
        return self.sig.process(data)

    @classmethod
    def Load(cls, data):
        if cls.__instance == None:
            cls.__instance = SignalLoader()
        return cls.__instance.process(data)


def load_signal(data):
    return SignalLoader.Load(data)

class FeatureExtractor(object):
    __instance = None
    def __init__(self, max_workers):
        self.executor = ProcessPoolExecutor(max_workers=max_workers)

    def extract(self, data):
        sig = load_signal(data)
        return self.executor.submit(madmom_feature_1024_3, sig)).result()
```

**Comment by laplaceyang** [comment]
seem same problem with [this](https://stackoverflow.com/questions/26598109/preserve-custom-attributes-when-pickling-subclass-of-numpy-array)

**Comment by laplaceyang** [comment]
> seem same problem with [this](https://stackoverflow.com/questions/26598109/preserve-custom-attributes-when-pickling-subclass-of-numpy-array)

fixed, after add function in my code
```
def signal__reduce__(self):
    # Get the parent's __reduce__ tuple
    pickled_state = super(Signal, self).__reduce__()
    # Create our own tuple to pass to __setstate__, but append the __dict__ rather than individual members.
    new_state = pickled_state[2] + (self.__dict__,)
    # Return a tuple that replaces the parent's __setstate__ tuple with our own
    return (pickled_state[0], pickled_state[1], new_state)

def signal__setstate__(self, state):
    self.__dict__.update(state[-1])  # Update the internal dict from state
    # Call the parent's __setstate__ with the other tuple elements.
    super(Signal, self).__setstate__(state[0:-1])

Signal.__reduce__ = signal__reduce__
Signal.__setstate__ = signal__setstate__
```

**Comment by superbock** [comment]
Thanks for reproducing this and providing a fix! See #455 for the PR.
